langlib/lang.array/src/main/java/org/ballerinalang/langlib/array/Next.java

```diff
@@ -49,16 +60,35 @@
     //TODO: refactor hard coded values
     public static Object next(Strand strand, ObjectValue m) {
         IteratorValue arrIterator = (IteratorValue) m.getNativeData("&iterator&");
-
+        ArrayValue arr = (ArrayValue) m.get(new BmpStringValue("m"));
         if (arrIterator == null) {
-            arrIterator = ((ArrayValue) m.get("m")).getIterator();
+            arrIterator = arr.getIterator();
             m.addNativeData("&iterator&", arrIterator);
         }
 
         if (arrIterator.hasNext()) {
             Object element =  arrIterator.next();
-            BFunctionType nextFuncType = m.getType().getAttachedFunctions()[0].type;
-            BRecordType recordType = (BRecordType) ((BUnionType) nextFuncType.retType).getMemberTypes().get(0);
+            Map<String, BField> fields = new HashMap<>();
+            BType type;
+            int typeFlags = 0;
+            if (arr.getType().getTag() == TypeTags.ARRAY_TAG) {
+                type = arr.getElementType();
+            } else {
+                BTupleType tupleType = (BTupleType) arr.getType();
+                LinkedHashSet<BType> types = new LinkedHashSet<>(tupleType.getTupleTypes());
+                if (tupleType.getRestType() != null) {
+                    types.add(tupleType.getRestType());
+                }
+                if (types.size() == 1) {
+                    type = types.iterator().next();
+                } else {
+                    type = new BUnionType(new ArrayList<>(types));
+                }
+            }
+
+            fields.put("value", new BField(type, "value", Flags.PUBLIC + Flags.REQUIRED));
+            BRecordType recordType = new BRecordType("$$returnType$$", null, 0, fields, null, true,
```

Comment:
With this approach, every time `next()` is called all these checks are done and a new type is created right? Can't we refactor this logic to maybe have the type be part of the iterator or at least add it as native data?

Reply:
Can we make the type be part of the iterator, because the type is always a typeParam?